pkg/dns/dns.go

```diff
@@ -404,40 +256,51 @@ func getMinTTL(m *dns.Msg, refreshRate time.Duration) time.Duration {
 	return minTTL
 }
 
-// Get domain name and refreshrate from cluster, and also store cluster and port in the return addresses for later use
-func getPendingResolveDomain(clusters []*clusterv3.Cluster) map[string]*pendingResolveDomain {
-	domains := make(map[string]*pendingResolveDomain)
+func (r *DNSResolver) GetAllCachedDomains() []string {
+	r.RLock()
+	defer r.RUnlock()
+	out := make([]string, 0, len(r.cache))
+	for domain := range r.cache {
+		out = append(out, domain)
+	}
+	return out
+}
+
+func (r *DNSResolver) GetOneDomainFromCache(domain string) ([]string, bool) {
+	r.Lock()
+	addresses, ok := r.cache[domain]
+	r.Unlock()
+	return addresses.Addresses, ok
+}
+
+func (r *DNSResolver) GetAddressesFromCache(domains map[string]struct{}) map[string]*DomainCacheEntry {
+	r.Lock()
+	defer r.Unlock()
 
-	for _, cluster := range clusters {
-		if cluster.LoadAssignment == nil {
-			continue
+	alreadyResolveDomains := make(map[string]*DomainCacheEntry)
+	for domain := range domains {
+		if v, ok := r.cache[domain]; ok {
+			alreadyResolveDomains[domain] = v
 		}
+	}
+	return alreadyResolveDomains
+}
 
-		for _, e := range cluster.LoadAssignment.Endpoints {
-			for _, le := range e.LbEndpoints {
-				socketAddr, ok := le.GetEndpoint().GetAddress().GetAddress().(*v3.Address_SocketAddress)
-				if !ok {
-					continue
-				}
-				address := socketAddr.SocketAddress.Address
-				if _, err := netip.ParseAddr(address); err == nil {
-					// This is an ip address
-					continue
-				}
-
-				if v, ok := domains[address]; ok {
-					v.clusters = append(v.clusters, cluster)
-				} else {
-					domainWithRefreshRate := &pendingResolveDomain{
-						domainName:  address,
-						clusters:    []*clusterv3.Cluster{cluster},
-						refreshRate: cluster.GetDnsRefreshRate().AsDuration(),
-					}
-					domains[address] = domainWithRefreshRate
-				}
-			}
+func (r *DNSResolver) RemoveUnwatchDomain(domains map[string]struct{}) {
+	r.Lock()
+	defer r.Unlock()
+
+	for domain := range r.cache {
+		if _, ok := domains[domain]; ok {
+			continue
 		}
+		delete(r.cache, domain)
 	}
+}
 
-	return domains
+func (r *DNSResolver) AddDomainIntoRefreshQueue(info *DomainInfo, time time.Duration) {
```

Comment:
It seems this is called by upper level controller, so prefer renaming to, e.g `EnqueDnsDomain`